impl/src/main/java/org/ehcache/impl/internal/store/heap/OnHeapStore.java

```diff
@@ -263,36 +263,29 @@ public OnHeapStore(final Configuration<K, V> config, final TimeSource timeSource
     final StoreEventSink<K, V> eventSink = storeEventDispatcher.eventSink();
 
     try {
-      final AtomicReference<OnHeapValueHolder<V>> expiredValue = new AtomicReference<OnHeapValueHolder<V>>(null);
+      OnHeapValueHolder<V> mapping = map.get(key);
 
-      OnHeapValueHolder<V> result = map.computeIfPresent(key, new BiFunction<K, OnHeapValueHolder<V>, OnHeapValueHolder<V>>() {
-        @Override
-        public OnHeapValueHolder<V> apply(K mappedKey, OnHeapValueHolder<V> mappedValue) {
-          final long now = timeSource.getTimeMillis();
-
-          if (mappedValue.isExpired(now, TimeUnit.MILLISECONDS)) {
-            onExpiration(mappedKey, mappedValue, eventSink);
-            expiredValue.set(mappedValue);
-            return null;
-          }
-
-          if (updateAccess) {
-            return setAccessTimeAndExpiryThenReturnMapping(key, mappedValue, now, eventSink);
-          }
+      if (mapping == null) {
+        getObserver.end(StoreOperationOutcomes.GetOutcome.MISS);
+        storeEventDispatcher.releaseEventSink(eventSink);
+        return null;
+      }
 
-          return mappedValue;
-        }
-      });
-      storeEventDispatcher.releaseEventSink(eventSink);
-      if (result == null) {
+      if (mapping.isExpired(timeSource.getTimeMillis(), TimeUnit.MILLISECONDS)) {
+        expireMapping(key, mapping);
         getObserver.end(StoreOperationOutcomes.GetOutcome.MISS);
-        if (expiredValue.get() != null) {
-          decrementCurrentUsageInBytesIfRequired(expiredValue.get().size());
-        }
-      } else {
+        storeEventDispatcher.releaseEventSink(eventSink);
+        return null;
+      }
+
+      if (updateAccess) {
+        mapping = setAccessTimeAndExpiryThenReturnMapping(key, mapping, timeSource.getTimeMillis(), eventSink);
```

Comment:
This is the only place we need the event sink, which means it should be created and released around this call.